timApp/static/scripts/tim/gamification/badge/badge-creator.component.ts

```diff
@@ -0,0 +1,776 @@
+            Validators.required,
+            this.disallowGrayColor,
+        ]),
+        shape: new FormControl("hexagon", [Validators.required]),
+        context_group: new FormControl("", [Validators.required]),
+    });
+
+    // Saves newly created badge
+    newBadge: any = null;
+    async onSubmit() {
+        if (this.badgeForm.valid) {
+            this.newBadge = this.badgeForm.value;
+            const response = toPromise(
+                this.http.post<{ok: boolean}>("/create_badge", {
+                    context_group: this.selectedContextGroup,
+                    title: this.newBadge.title,
+                    color: this.newBadge.color,
+                    shape: this.newBadge.shape,
+                    image: this.newBadge.image.toString(),
+                    description: this.newBadge.description,
+                })
+            );
+            // console.log("contextgroup: ", this.selectedContextGroup);
+            const result = await response;
+            // console.log("contextgroup: ", this.selectedContextGroup);
+            if (result.ok) {
+                while (this.all_badges.length > 0) {
+                    this.all_badges.pop();
+                }
+                await this.getBadges();
+                this.emptyForm();
+
+                this.badgeService.triggerUpdateBadgeList();
+                this.badgeFormShowing = false;
+                this.centerToComponent();
+            }
+            if (!result.ok) {
+                if (result.result.error.error == undefined) {
+                    this.badgeService.showError(
+                        this.alerts,
+                        {
+                            data: {
+                                error: "Unexpected error. Check your internet connection.",
+                            },
+                        },
+                        "danger"
+                    );
+                    return;
+                }
+                this.badgeService.showError(
+                    this.alerts,
+                    {data: {error: result.result.error.error}},
+                    "danger"
+                );
+                return;
+            }
+        }
+    }
+
+    // When cancelled, form information is cleared
+    async onCancel() {
+        this.emptyForm();
+        await this.getBadges();
+        this.badgeFormShowing = false;
+        this.resetForm();
+        setTimeout(() => {
+            this.centerToComponent();
+        }, 100);
+    }
+
+    // Get all badges depending on if context group is selected. Fails if not teacher.
+    private async getBadges() {
+        let response;
+        if (this.selectedContextGroup) {
+            response = toPromise(
+                this.http.get<[]>(
+                    `/all_badges_in_context/${this.selectedContextGroup}`
+                )
+            );
+        } else {
+            response = toPromise(this.http.get<[]>("/all_badges"));
```

Comment:
Pitäiskö kaikki badget näyttää yhtenä pötkönä vai eroteltuina context-group -ryhmittäin? Pitäs ainakin jatkokehitykseen kirjata että pitää miettiä